keps/0027-pod-restart-controls.md

```diff
@@ -0,0 +1,189 @@
+
+A new flag introduced by this KEP will allow an operator developer to control when a changed parameter will trigger a restart of pods from a deployment or stateful set.
+
+### Goals
+
+Make it possible for an operator developer to specify when a parameter will restart all pods from a stateful set or a deployment.
+
+### Non-Goals
+
+- Preventing Pod restart where plain k8s would do it: If the spec template for a stateful set is changed, Kubernetes will still restart all pods automatically - there is no way to prevent that.
+- Calculation of used config maps, secrets or other resources used by a pod: The automatic pod restart is currently used to force pods to re-read config maps on a parameter update. When a user disables the pod restart for a parameter, and this parameter updates a config map or another dependency of the pod, the change might not be applied in the pod until the next restart.
+- Restarting a stuck Pod (A step that cannot be executed, a pod that cannot be deployed because of various reasons, a pod that crash loops, etc.)
+
+## Proposal 1 - Parameter Groups
+
+Allow the operator developer to define groups of parameters that can then be used to trigger a restart of pods from a deployment or stateful set.
+
+Add a list of `groups` that each parameter is part of:
+
+```yaml
+  - name: CONFIG_VALUE
+    description: "A parameter that changes a config map which is used by a stateful set."
+    default: "3"
+    groups:
+      - MAIN_GROUP
+  - name: SOME_OTHER_PARAMETER
+    description: "A parameter that changes something that is used by two stateful sets."
+    default: "3"
+    groups:
+      - MAIN_GROUP
+      - ADDITIONAL_GROUP
+```
+
+KUDO then calculates the hash of all parameters in a group. This hash can then be used instead of the the `last-plan-execution-uid` in a deployment or stateful set to trigger the reload of the pods:
+
+```yaml
+spec:
+  template:
+    metadata:
+      annotations:
+        config-values-hash: {{ .ParamGroupHashses.MAIN_GROUP }}
+```
+
+When a parameter in the group changes, the hash will change, which in turn would then trigger the restart of the pods.
+
+This proposal would remove the `last-plan-execution-uid` and automatic pod restarts. Configuration for pod restarts would be fully in the responsibility of the operator developer.  
+
+There will be a default `ALL` group that all parameters belong to. For an operator developer to restore the current behavior, the would only need to add the following annotation to the spec template:
+
+```yaml
+spec:
+  template:
+    metadata:
+      annotations:
+        config-values-hash: {{ .ParamGroupHashses.ALL }}
+```
+
+## Proposal 2 - Dependency Graph
+
+KUDO will analyze the operator and build a dependency graph of resources:
```

Comment:
what about instead of analyzing, providing developers a way to register the dependencies?
```yaml
spec:
  template:
    metadata:
      annotations:
        health-check-cm: {{ .TemplatesHash.healthcheck.yaml }}
        bootstrap-script: {{ .TemplatesHash.bootstrap.sh.yaml }}
```

on each update, we just update the hashes where they are used and make it easier the "dependency" approach.
WDYT?

Reply:
That would be easier on us, but prone to getting out of sync with real dependencies. Analyzing actual ones means we have a single source of truth.

Reply:
Hmmm, i don't think the analysis would be too hard in these cases, so I don't think that part makes it more compelling.

If we were to go with this proposal, we'd at least have to replace any dots and special chars:
```
 - bootstrap-script: {{ .TemplatesHash.bootstrap-sh-yaml }}
```
This might get more annoying if we at some point allow resources to be in subdirectories.

Pro:
- It would remove the requirement to have an annotation in a configmap/secret when that resource should *not* be included in the hash
- It would make very clear which resources trigger a pod restart

Con:
- It would require the operator developers to add new resources to the annotations. 
- To replicate the current behavior, an operator developer would have to add all used resources in annotations


I'm very undecided. My gut feeling likes the dependency variant better.